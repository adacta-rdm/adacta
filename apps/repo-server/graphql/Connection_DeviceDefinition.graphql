# @generated

type Connection_DeviceDefinition {
	edges: [Edge_DeviceDefinition!]!
	pageInfo: PageInfo!
}
