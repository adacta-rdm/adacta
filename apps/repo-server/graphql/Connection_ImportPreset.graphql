# @generated

type Connection_ImportPreset {
	edges: [Edge_ImportPreset!]!
	pageInfo: PageInfo!
}
