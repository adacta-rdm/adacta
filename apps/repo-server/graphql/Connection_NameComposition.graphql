# @generated

type Connection_NameComposition {
	edges: [Edge_NameComposition!]!
	pageInfo: PageInfo!
}
