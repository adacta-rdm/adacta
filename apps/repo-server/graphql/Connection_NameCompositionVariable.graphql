# @generated

type Connection_NameCompositionVariable {
	edges: [Edge_NameCompositionVariable!]!
	pageInfo: PageInfo!
}
