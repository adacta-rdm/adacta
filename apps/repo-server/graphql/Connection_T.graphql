interface Connection_T {
	edges: [Edge_T!]!
	pageInfo: PageInfo!
}
