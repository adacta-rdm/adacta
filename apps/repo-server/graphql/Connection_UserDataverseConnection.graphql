# @generated

type Connection_UserDataverseConnection {
	edges: [Edge_UserDataverseConnection!]!
	pageInfo: PageInfo!
}
