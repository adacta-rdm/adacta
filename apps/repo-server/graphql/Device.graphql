type Device implements Node & SupportsUsageAsProperty & SupportsNotes & HasMetadata & HasImageResource & HasProjects {
	id: ID!
	name: String!
	displayName: String!
	metadata: Metadata!

	"""
	A unique identifier for this device. It is used in the URL of the device page.
	"""
	shortId: String

	definition: DeviceDefinition!

	definitions: [DeviceDefinitionGraphElement!]!
	"""
	Returns the combined specifications from the device and the inheritance tree
	"""
	specifications(names: [String!]): [Specification!]!
	properties(timestamp: DateTime): [Property!]!

	imageResource: [ResourceImage!]!
	setupDescription(time: DateTime): [SetupDescription!]!

	usagesAsProperty(
		timeFrame: TimeFrameInput
		time: DateTime
		includeOverlaps: Boolean
	): [Property!]!

	# The parent device at the given time
	parent(timestamp: DateTime): Device

	samples: [SampleUsage!]!

	usageInResource: [ResourceTimed]!

	"""
	Returns a flat list of all child devices. The list corresponds to the components this device is
	made of in the specified timespan. A value of `null` for the `end` argument means "up until
	now".
	"""
	components(
		timeFrame: TimeFrameInput
		time: DateTime
		includeOverlaps: Boolean
	): [ComponentWithPathAndTime!]!
	# Can be undefined if the device is not installed anywhere
	topLevelDevice(timestamp: DateTime): Device

	"""
	Returns the components that are available for use in the given time frame.
	This resolver is used to find possible components to be used as sub-components of a device.
	A timeframe is required as the device must be unused for the entire duration.

	The property with the ID `ignoreProperty` should be ignored when checking for conflicts, because
	it is the property that is being updated (and therefore won't cause a conflict with itself).
	"""
	freeComponents(begin: DateTime!, end: DateTime, ignoreProperty: ID): [PropertyValue!]!

	componentsInSlot(path: [String!]!): [ComponentWithPathAndTime!]!

	notes: NoteConnection!

	projects(first: Int, after: String): ProjectConnection!
}
