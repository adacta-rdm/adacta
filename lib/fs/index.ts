/**
 * @file Automatically generated by barrelsby.
 */

export * from "./exists";
export * from "./isDir";
export * from "./mkdirTmp";
export * from "./mkdirp";
export * from "./readFile";
export * from "./readJSONFile";
export * from "./readUTF8File";
export * from "./readdir";
export * from "./rmrf";
export * from "./stat";
export * from "./writeFile";
export * from "./private/normalizePath";
