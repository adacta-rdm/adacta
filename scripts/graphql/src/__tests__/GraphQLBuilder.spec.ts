import { readFileSync } from "node:fs";
import { join } from "node:path";

import { print } from "graphql";
import { describe, test, expect } from "vitest";

import { GraphQLBuilder } from "../GraphQLBuilder";

const queries = readFileSync(join(__dirname, "./typeDefs/queries.graphql"), "utf-8");
const types = readFileSync(join(__dirname, "./typeDefs/types.graphql"), "utf-8");

describe("GraphQLBuilder", () => {
	describe("#operations()", () => {
		test("returns operations", () => {
			const builder = new GraphQLBuilder([queries]);

			expect(builder.operations()).toMatchSnapshot();
		});
	});

	describe("#typeDefinitionNode()", () => {
		function setup(typeDefs = [types]) {
			const builder = new GraphQLBuilder(typeDefs);

			return {
				builder,
				typeDef: (typeName: string) => print(builder.typeDefinitionNode(typeName)),
			};
		}

		test("returns type def for defined type", () => {
			const { typeDef } = setup();

			const book = typeDef("Book");

			expect(book).toBe(
				[
					"type Book {", //
					"  id: ID!",
					"  wrappedString: Wrapper_String!",
					"}",
				].join("\n")
			);
		});

		test("returns the concrete type def for a template type", () => {
			const { typeDef } = setup();

			const wrapper = typeDef("Wrapper_String");

			expect(wrapper).toBe(
				[
					"type Wrapper_String {", //
					"  value: String!",
					"}",
				].join("\n")
			);
		});

		test("returns the concrete type def for a template type and resolves template type fields", () => {
			const { typeDef } = setup();

			const wrapper = typeDef("Connection_String");

			expect(wrapper).toBe(
				[
					"type Connection_String {", //
					"  edges: [Edge_String!]!",
					"}",
				].join("\n")
			);
		});

		test("returns special T type", () => {
			const { typeDef } = setup();

			const wrapper = typeDef("T");

			expect(wrapper).toBe(
				[
					"scalar T", //
				].join("\n")
			);
		});

		test("returns scalar type", () => {
			const { typeDef } = setup();

			const scalar = typeDef("Scalar1");

			expect(scalar).toBe("scalar Scalar1");
		});

		test("returns enum type", () => {
			const { typeDef } = setup();

			const enumTypeDef = typeDef("Enum1");

			expect(enumTypeDef).toBe(
				[
					//
					"enum Enum1 {",
					"  A",
					"  B",
					"}",
				].join("\n")
			);
		});

		test("template input type", () => {
			const { typeDef } = setup();

			//
			// input Input1 {
			// 	prop1: Float!
			// 	prop2: String!
			// 	prop3: Int!
			// }
			//
			// input Input2_T {
			// 	prop1: Partial_T!
			// }

			const inputTypeDef = typeDef("Input2_Input1");

			expect(inputTypeDef).toBe(
				[
					//
					"input Input2_Input1 {",
					"  prop1: Partial_Input1",
					"}",
				].join("\n")
			);
		});

		test("returns union type", () => {
			const { typeDef } = setup();

			const union = typeDef("Union1");

			expect(union).toBe("union Union1 = Wrapper_String | Wrapper_Int");
		});

		describe("Partial_T", () => {
			test("makes all fields of type nullable", () => {
				const { typeDef } = setup([
					[
						"type Query {",
						"  books: Connection_Book!",
						"  book(id: ID!): Book",
						"  list: [Book!]!",
						"}",
					].join("\n"),
				]);

				const partialQuery = typeDef("Partial_Query");

				expect(partialQuery).toBe(
					[
						"type Partial_Query {", //
						"  books: Connection_Book",
						"  book(id: ID!): Book",
						"  list: [Book!]",
						"}",
					].join("\n")
				);
			});

			test("makes all fields of interface nullable", () => {
				const { typeDef } = setup();

				const partialInterface = typeDef("Partial_Interface1");

				expect(partialInterface).toBe(
					[
						"interface Partial_Interface1 {", //
						"  prop1: Float",
						"  prop2: String",
						"  prop3: Int",
						"}",
					].join("\n")
				);
			});

			test("makes all fields of input nullable", () => {
				const { typeDef } = setup();

				const partialInput = typeDef("Partial_Input1");

				expect(partialInput).toBe(
					[
						"input Partial_Input1 {", //
						"  prop1: Float",
						"  prop2: String",
						"  prop3: Int",
						"}",
					].join("\n")
				);
			});

			test("returns special Partial_T type", () => {
				const { typeDef } = setup();

				const wrapper = typeDef("Partial_T");

				expect(wrapper).toBe(
					[
						"scalar Partial_T", //
					].join("\n")
				);
			});
		});
	});

	describe("#generateTypeDefs()", () => {
		test("contains all instantiated template types", () => {
			const builder = new GraphQLBuilder([types]);

			const typeDefs = [...builder.generateTypeDefs().keys()];
			expect(typeDefs).toHaveLength(7);
			expect(typeDefs).toContain("Connection_Book");
			expect(typeDefs).toContain("Edge_Book");
			expect(typeDefs).toContain("Wrapper_String");
			expect(typeDefs).toContain("Wrapper_Int");
			expect(typeDefs).toContain("Partial_Input1");
			expect(typeDefs).toContain("Partial_T");
			expect(typeDefs).toContain("T");
		});

		test("recursive type", () => {
			const builder = new GraphQLBuilder(
				[
					"type RecursiveQuery {",
					"  prop1(id: ID!): RecursiveQuery!",
					"  prop2(id: ID!): Wrapper_RecursiveQuery",
					"}",
					//
					"type Wrapper_T {",
					"  value: T",
					"}",
				].join("\n")
			);

			const typeDefs = [...builder.generateTypeDefs().keys()];

			expect(typeDefs).toHaveLength(2);
			expect(typeDefs).toContain("Wrapper_RecursiveQuery");
			expect(typeDefs).toContain("T");
		});
	});

	describe("#bundle()", () => {
		test("include only referenced types", () => {
			const builder = new GraphQLBuilder([types]);

			const document = builder.bundle();

			const typeDefs = print(document);

			expect(document.definitions).toHaveLength(7);

			expect(typeDefs).toContain("schema");
			expect(typeDefs).toContain("type Query");
			expect(typeDefs).toContain("type Connection_Book");
			expect(typeDefs).toContain("type Edge_Book");
			expect(typeDefs).toContain("type Book");
			expect(typeDefs).toContain("type Wrapper_String");
			expect(typeDefs).toContain("input Partial_Input1");
			expect(typeDefs).not.toContain("type Connection_T");
			expect(typeDefs).not.toContain("type Edge_T");
			expect(typeDefs).not.toContain("type Wrapper_T");
			expect(typeDefs).not.toContain("type Unused");
		});

		test("include only referenced types - explicit entry point", () => {
			const builder = new GraphQLBuilder([types]);

			const document = builder.bundle("Type1");

			const typeDefs = print(document);

			expect(document.definitions).toHaveLength(1);

			expect(typeDefs).toContain("type Type1");
			expect(typeDefs).not.toContain("schema");
			expect(typeDefs).not.toContain("type Query");
			expect(typeDefs).not.toContain("type Connection_Book");
			expect(typeDefs).not.toContain("type Edge_Book");
			expect(typeDefs).not.toContain("type Book");
			expect(typeDefs).not.toContain("type Wrapper_String");
			expect(typeDefs).not.toContain("type Connection_T");
			expect(typeDefs).not.toContain("type Edge_T");
			expect(typeDefs).not.toContain("type Wrapper_T");
		});

		test("recursive type", () => {
			const builder = new GraphQLBuilder(
				[
					//
					"type RecursiveQuery {",
					"  prop(id: ID!): RecursiveQuery!",
					"}",
				].join("\n")
			);

			const document = builder.bundle("RecursiveQuery");

			const typeDefs = print(document);

			expect(document.definitions).toHaveLength(1);

			expect(typeDefs).toContain("type RecursiveQuery");
		});

		test("union type", () => {
			const builder = new GraphQLBuilder([types]);

			const document = builder.bundle("Union1");

			const typeDefs = print(document);

			expect(document.definitions).toHaveLength(3);

			expect(typeDefs).toContain("union Union1 = Wrapper_String | Wrapper_Int");
			expect(typeDefs).toContain("type Wrapper_String");
			expect(typeDefs).toContain("type Wrapper_Int");
		});

		test("interface type", () => {
			// Interfaces should be resolved to their implementing types and included in the bundle.
			const builder = new GraphQLBuilder(
				[
					//
					"type Query {",
					"  node(id: ID!): Node",
					"}",
					//
					"interface Node {",
					"  id: ID!",
					"}",
					//
					"type Book implements Node {",
					"  id: ID!",
					"  title: String!",
					"}",
					//
					"type Author implements Node {",
					"  id: ID!",
					"  name: String!",
					"}",
				].join("\n")
			);

			const document = builder.bundle("Query");

			const typeDefs = print(document);

			expect(document.definitions).toHaveLength(4);

			expect(typeDefs).toContain("type Query");
			expect(typeDefs).toContain("interface Node");
			expect(typeDefs).toContain("type Book");
			expect(typeDefs).toContain("type Author");
		});

		test("directives type", () => {
			// Directive definitions should be included in the bundle.
			const builder = new GraphQLBuilder(
				[
					"type Query {",
					"  node(id: ID!): Node",
					"}",
					//
					"interface Node {",
					"  id: ID!",
					"}",
					//
					"type Book implements Node {",
					"  id: ID!",
					"  title: String! @myDirective1",
					"}",
					//
					"directive @myDirective1 on FIELD_DEFINITION",
					"directive @myDirective2 on FIELD_DEFINITION",
				].join("\n")
			);

			const document = builder.bundle("Query");

			const typeDefs = print(document);

			// expect(document.definitions).toHaveLength(4);

			expect(typeDefs).toContain("directive @myDirective1");
			expect(typeDefs).not.toContain("directive @myDirective2");
		});
	});
});
