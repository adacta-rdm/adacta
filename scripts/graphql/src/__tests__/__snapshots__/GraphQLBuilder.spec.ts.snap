// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`GraphQLBuilder > #operations() > returns operations 1`] = `
{
  "definitions": [
    {
      "directives": [],
      "kind": "OperationDefinition",
      "loc": {
        "end": 88,
        "start": 0,
      },
      "name": {
        "kind": "Name",
        "loc": {
          "end": 12,
          "start": 6,
        },
        "value": "Query1",
      },
      "operation": "query",
      "selectionSet": {
        "kind": "SelectionSet",
        "loc": {
          "end": 88,
          "start": 26,
        },
        "selections": [
          {
            "alias": undefined,
            "arguments": [
              {
                "kind": "Argument",
                "loc": {
                  "end": 44,
                  "start": 34,
                },
                "name": {
                  "kind": "Name",
                  "loc": {
                    "end": 36,
                    "start": 34,
                  },
                  "value": "id",
                },
                "value": {
                  "kind": "Variable",
                  "loc": {
                    "end": 44,
                    "start": 38,
                  },
                  "name": {
                    "kind": "Name",
                    "loc": {
                      "end": 44,
                      "start": 39,
                    },
                    "value": "input",
                  },
                },
              },
            ],
            "directives": [],
            "kind": "Field",
            "loc": {
              "end": 86,
              "start": 29,
            },
            "name": {
              "kind": "Name",
              "loc": {
                "end": 33,
                "start": 29,
              },
              "value": "book",
            },
            "selectionSet": {
              "kind": "SelectionSet",
              "loc": {
                "end": 86,
                "start": 46,
              },
              "selections": [
                {
                  "alias": undefined,
                  "arguments": [],
                  "directives": [],
                  "kind": "Field",
                  "loc": {
                    "end": 52,
                    "start": 50,
                  },
                  "name": {
                    "kind": "Name",
                    "loc": {
                      "end": 52,
                      "start": 50,
                    },
                    "value": "id",
                  },
                  "selectionSet": undefined,
                },
                {
                  "alias": undefined,
                  "arguments": [],
                  "directives": [],
                  "kind": "Field",
                  "loc": {
                    "end": 83,
                    "start": 55,
                  },
                  "name": {
                    "kind": "Name",
                    "loc": {
                      "end": 68,
                      "start": 55,
                    },
                    "value": "wrappedString",
                  },
                  "selectionSet": {
                    "kind": "SelectionSet",
                    "loc": {
                      "end": 83,
                      "start": 69,
                    },
                    "selections": [
                      {
                        "alias": undefined,
                        "arguments": [],
                        "directives": [],
                        "kind": "Field",
                        "loc": {
                          "end": 79,
                          "start": 74,
                        },
                        "name": {
                          "kind": "Name",
                          "loc": {
                            "end": 79,
                            "start": 74,
                          },
                          "value": "value",
                        },
                        "selectionSet": undefined,
                      },
                    ],
                  },
                },
              ],
            },
          },
        ],
      },
      "variableDefinitions": [
        {
          "defaultValue": undefined,
          "directives": [],
          "kind": "VariableDefinition",
          "loc": {
            "end": 24,
            "start": 13,
          },
          "type": {
            "kind": "NonNullType",
            "loc": {
              "end": 24,
              "start": 21,
            },
            "type": {
              "kind": "NamedType",
              "loc": {
                "end": 23,
                "start": 21,
              },
              "name": {
                "kind": "Name",
                "loc": {
                  "end": 23,
                  "start": 21,
                },
                "value": "ID",
              },
            },
          },
          "variable": {
            "kind": "Variable",
            "loc": {
              "end": 19,
              "start": 13,
            },
            "name": {
              "kind": "Name",
              "loc": {
                "end": 19,
                "start": 14,
              },
              "value": "input",
            },
          },
        },
      ],
    },
  ],
  "kind": "Document",
}
`;
